src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java

```diff
@@ -256,11 +422,92 @@ public Void visitMethod(MethodTree tree, AnnotatedTypeMirror p) {
       ExecutableElement element = TreeUtils.elementFromDeclaration(tree);
       if (ElementUtils.hasAnnotation(element, Provides.class.getName())) {
         // TODO: Put fully qualified names of classes in knownBindings
-        knownBindings.put(tree.getReturnType().toString(), p.getUnderlyingType().toString());
+        knownBindings.put(
+            tree.getReturnType().toString(),
+            KnownBindingsValue.builder().className(p.getUnderlyingType().toString()).build());
       }
 
       printKnownBindings();
       return super.visitMethod(tree, p);
     }
   }
+
+  @Override
+  protected QualifierHierarchy createQualifierHierarchy() {
+    return new DependencyInjectionQualifierHierarchy(
+        this.getSupportedTypeQualifiers(), this.elements);
+  }
+
+  protected class DependencyInjectionQualifierHierarchy extends AccumulationQualifierHierarchy {
+
+    protected DependencyInjectionQualifierHierarchy(
+        Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+      super(qualifierClasses, elements);
+    }
+
+    @Override
+    public boolean isSubtype(final AnnotationMirror subAnno, final AnnotationMirror superAnno) {
+
+      if (AnnotationUtils.areSame(subAnno, top)) {
+        if (AnnotationUtils.areSame(superAnno, top)) {
```

Comment:
Rather than having this check within each of these other blocks, it would be simpler to just return true at the start of the method if `superAnno` is top.

Reply:
This makes sense because if `superAnno` is top then `subAnno` will always be a subtype of it.